File: Trees/ValidateBST.cpp
#include <climits>
// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  bool isValidBST(TreeNode *root) { return dfs(root, LONG_MIN, LONG_MAX); }

  bool dfs(TreeNode *root, long left, long right) {
    if (!root) {
      return true;
    }

    if (root->val > left && root->val < right) {
      return dfs(root->left, left, root->val) &&
             dfs(root->right, root->val, right);
    }
    return false;
  }
};

File: Blind75/Trees/ValidateBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(root, maxLeft, minRight):
            if not root:
                return True
            if not (root.val > maxLeft and root.val < minRight):
                return False

            return dfs(root.left, maxLeft, root.val) and dfs(
                root.right, root.val, minRight
            )

        return dfs(root, float("-inf"), float("inf"))
